File: app/src/main/java/com/adsamcik/signalcollector/Fragments/FragmentStats.java
package com.adsamcik.signalcollector.fragments;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.content.ContextCompat;
import android.support.v4.widget.SwipeRefreshLayout;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.ScrollView;

import com.adsamcik.signalcollector.utility.Assist;
import com.adsamcik.signalcollector.utility.Failure;
import com.adsamcik.signalcollector.utility.NetworkLoader;
import com.adsamcik.signalcollector.utility.Preferences;
import com.adsamcik.signalcollector.R;
import com.adsamcik.signalcollector.activities.RecentUploadsActivity;
import com.adsamcik.signalcollector.utility.DataStore;
import com.adsamcik.signalcollector.utility.Network;
import com.adsamcik.signalcollector.utility.Signin;
import com.adsamcik.signalcollector.utility.SnackMaker;
import com.adsamcik.signalcollector.utility.Table;
import com.adsamcik.signalcollector.data.UploadStats;
import com.adsamcik.signalcollector.data.Stat;
import com.adsamcik.signalcollector.data.StatData;
import com.adsamcik.signalcollector.data.StatDay;
import com.adsamcik.signalcollector.interfaces.ITabFragment;

import java.util.ArrayList;

public class FragmentStats extends Fragment implements ITabFragment {
	private Table weeklyStats;
	private Table lastUpload;
	private final ArrayList<Table> publicStats = new ArrayList<>();
	private final ArrayList<Table> userStats = new ArrayList<>();
	private View view;

	private SwipeRefreshLayout refreshLayout;

	//todo add user stats

	@Override
	public void onCreate(@Nullable Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		Context context = getContext();
		weeklyStats = new Table(context, 4, false, ContextCompat.getColor(context, R.color.textPrimary));
	}

	@Override
	public void onDestroyView() {
		((LinearLayout) view.findViewById(R.id.statsLayout)).removeAllViews();
		publicStats.clear();
		userStats.clear();
		super.onDestroyView();
	}

	@Nullable
	@Override
	public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
		view = inflater.inflate(R.layout.fragment_stats, container, false);
		Resources r = getResources();

		boolean hasRecentUpload = false;

		UploadStats us = DataStore.loadLastObjectJsonArrayAppend(DataStore.RECENT_UPLOADS_FILE, UploadStats.class);
		if (us != null && Assist.getAgeInDays(us.time) < 30) {
			lastUpload = RecentUploadsActivity.GenerateTableForUploadStat(us, (LinearLayout) view.findViewById(R.id.statsLayout), getContext(), getResources().getString(R.string.most_recent_upload));
			lastUpload.addButton(getString(R.string.more_uploads), v -> {
				Intent intent = new Intent(getContext(), RecentUploadsActivity.class);
				startActivity(intent);
			});
			hasRecentUpload = true;
		} else {
			if (lastUpload != null)
				lastUpload.destroy(getActivity());
		}

		new Thread(DataStore::removeOldRecentUploads).start();

		Preferences.checkStatsDay(getActivity());

		weeklyStats.clear();
		weeklyStats.addTitle(r.getString(R.string.stats_weekly_title));
		StatDay weekStats = Preferences.countStats(getActivity());
		weeklyStats.addRow().addData(r.getString(R.string.stats_weekly_minutes), String.valueOf(weekStats.getMinutes()));
		weeklyStats.addRow().addData(r.getString(R.string.stats_weekly_uploaded), Assist.humanReadableByteCount(weekStats.getUploaded(), true));
		weeklyStats.addRow().addData(r.getString(R.string.stats_weekly_collected_location), String.valueOf(weekStats.getLocations()));
		weeklyStats.addRow().addData(r.getString(R.string.stats_weekly_collected_wifi), String.valueOf(weekStats.getWifi()));
		weeklyStats.addRow().addData(r.getString(R.string.stats_weekly_collected_cell), String.valueOf(weekStats.getCell()));
		weeklyStats.addToViewGroup((LinearLayout) view.findViewById(R.id.statsLayout), hasRecentUpload ? 1 : 0, false, 0);

		updateStats();
		refreshLayout = (SwipeRefreshLayout) view.findViewById(R.id.statsSwipeRefresh);
		refreshLayout.setOnRefreshListener(this::updateStats);
		refreshLayout.setColorSchemeResources(R.color.colorPrimary);
		return view;
	}

	private void updateStats() {
		Activity activity = getActivity();
		final boolean isRefresh = refreshLayout != null && refreshLayout.isRefreshing();
		refreshingCount++;
		NetworkLoader.load(Network.URL_STATS, isRefresh ? 0 : Assist.DAY_IN_MINUTES, getContext(), Preferences.GENERAL_STATS, Stat[].class, (state, value) -> {
			refreshDone();
			final int initialIndex = ((ViewGroup) view).getChildCount();

			if (state.isSuccess())
				generateStats(value, publicStats, initialIndex, activity);
			else {
				generateStats(value, publicStats, initialIndex, activity);
				new SnackMaker(activity).showSnackbar(state.toString(activity));
			}
		});

		if (Signin.getToken(getActivity()) != null) {
			refreshingCount++;
			NetworkLoader.request(Network.URL_USER_STATS, isRefresh ? 0 : Assist.DAY_IN_MINUTES, getContext(), Preferences.USER_STATS, Stat[].class, (state, value) -> {
				refreshDone();
				final int initialIndex = 1 + (lastUpload == null ? 0 : 1);
				if (state.isSuccess())
					generateStats(value, userStats, initialIndex, activity);
				else {
					generateStats(value, userStats, initialIndex, activity);
					new SnackMaker(activity).showSnackbar(state.toString(activity));
				}
			});
		}
	}

	private int refreshingCount = 0;

	private void refreshDone() {
		if (--refreshingCount == 0)
			if (refreshLayout != null && refreshLayout.isRefreshing())
				getActivity().runOnUiThread(() -> refreshLayout.setRefreshing(false));
	}

	private void generateStats(Stat[] stats, ArrayList<Table> items, int insertAt, Activity activity) {
		if (stats == null)
			return;

		if (items != null) {
			for (Table t : items) {
				t.destroy(activity);
			}
		}
		activity.runOnUiThread(() -> generateStatsTable(stats, insertAt, items));
	}

	/**
	 * Generates tables from list of stats
	 *
	 * @param stats stats
	 * @param items array to which items will be added
	 * @return returns passed items array
	 */
	private ArrayList<Table> generateStatsTable(Stat[] stats, int insertAt, ArrayList<Table> items) {
		Context c = getContext();
		LinearLayout ll = (LinearLayout) view.findViewById(R.id.statsLayout);
		int color = ContextCompat.getColor(c, R.color.textPrimary);
		for (int i = 0; i < stats.length; i++) {
			Stat s = stats[i];
			if (s.data != null) {
				Table table = new Table(c, s.data.size(), s.showPosition, color);
				table.addTitle(s.name);
				for (int y = 0; y < s.data.size(); y++) {
					StatData sd = s.data.get(y);
					table.addRow().addData(sd.id, sd.value);
				}
				table.addToViewGroup(ll, insertAt + i, true, (i + 1) * 150);
				items.add(table);
			}
		}
		return items;
	}

	@Override
	public Failure<String> onEnter(FragmentActivity activity, FloatingActionButton fabOne, FloatingActionButton fabTwo) {
		//todo check if up to date
		return new Failure<>();
	}

	@Override
	public void onLeave() {
		if (refreshLayout != null && refreshLayout.isRefreshing())
			refreshLayout.setRefreshing(false);
	}

	@Override
	public void onPermissionResponse(int requestCode, boolean success) {

	}

	@Override
	public void onHomeAction() {
		Assist.verticalSmoothScrollTo((ScrollView) view.findViewById(R.id.statsLayout).getParent(), 0, 500);
	}

}

File: app/src/main/java/com/adsamcik/signalcollector/fragments/FragmentStats.kt
package com.adsamcik.signalcollector.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ListView
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.adsamcik.draggable.IOnDemandView
import com.adsamcik.signalcollector.R
import com.adsamcik.signalcollector.adapters.ChangeTableAdapter
import com.adsamcik.signalcollector.data.*
import com.adsamcik.signalcollector.database.AppDatabase
import com.adsamcik.signalcollector.database.data.DatabaseLocation
import com.adsamcik.signalcollector.extensions.dpAsPx
import com.adsamcik.signalcollector.extensions.format
import com.adsamcik.signalcollector.uitools.ColorManager
import com.adsamcik.signalcollector.uitools.ColorSupervisor
import com.adsamcik.signalcollector.uitools.ColorView
import com.adsamcik.signalcollector.utility.Assist
import com.adsamcik.signalcollector.utility.Constants
import com.adsamcik.table.AppendBehaviors
import com.adsamcik.table.Table
import com.adsamcik.table.TableAdapter
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.util.*

class FragmentStats : Fragment(), IOnDemandView {
	private lateinit var fragmentView: View

	private var adapter: TableAdapter? = null

	private lateinit var swipeRefreshLayout: SwipeRefreshLayout

	private lateinit var colorManager: ColorManager

	override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
		val activity = activity!!
		colorManager = ColorSupervisor.createColorManager(activity)

		val fragmentView = inflater.inflate(R.layout.fragment_stats, container, false)


		adapter = ChangeTableAdapter(activity, CARD_LIST_MARGIN, activity.packageManager.getActivityInfo(activity.componentName, 0).themeResource)

		//weeklyStats.addToViewGroup(view.findViewById(R.id.statsLayout), hasRecentUpload ? 1 : 0, false, 0);

		swipeRefreshLayout = fragmentView.findViewById(R.id.swiperefresh_stats)
		swipeRefreshLayout.setOnRefreshListener { this.updateStats() }
		swipeRefreshLayout.setColorSchemeResources(R.color.color_primary)
		swipeRefreshLayout.setProgressViewOffset(true, 0, 40.dpAsPx)

		val listView = fragmentView!!.findViewById<ListView>(R.id.listview_stats)
		listView.setRecyclerListener { }
		listView.adapter = adapter
		updateStats()

		this.fragmentView = fragmentView

		colorManager.watchAdapterView(ColorView(listView, 1, recursive = true, rootIsBackground = true))

		return fragmentView
	}

	override fun onDestroyView() {
		ColorSupervisor.recycleColorManager(colorManager)
		super.onDestroyView()
	}

	private fun updateStats() {
		val activity = activity
		val appContext = activity!!.applicationContext
		val isRefresh = swipeRefreshLayout.isRefreshing

		adapter!!.clear()

		val r = activity.resources

		GlobalScope.launch(Dispatchers.Main, CoroutineStart.DEFAULT) {
			swipeRefreshLayout.isRefreshing = true
		}


		/*if (useMock) {
			generateMockData()
		} else {*/
		//refreshingCount = 2
		//new stat loading
		GlobalScope.launch {
			val database = AppDatabase.getAppDatabase(activity)
			val sessionDao = database.sessionDao()
			val wifiDao = database.wifiDao()
			val cellDao = database.cellDao()

			val calendar = Calendar.getInstance()

			val now = calendar.timeInMillis

			calendar.add(Calendar.WEEK_OF_YEAR, -1)
			val weekAgo = calendar.timeInMillis

			val sumSession = TrackingSession(now, now, 0, 0f, 0)
			var totalMinutes = 0.0
			sessionDao.getBetween(weekAgo, now).forEach {
				sumSession.mergeWith(it)
				val time = it.end - it.start
				totalMinutes += time.toDouble() / Constants.MINUTE_IN_MILLISECONDS.toDouble()
			}

			val weeklyStats = arrayOf(Stat(r.getString(R.string.stats_weekly_title), "", showPosition = false, data = listOf(
					StatData(r.getString(R.string.stats_weekly_minutes), totalMinutes.toString()),
					StatData(r.getString(R.string.stats_weekly_collected_location), sumSession.collections.toString()),
					StatData(r.getString(R.string.stats_weekly_steps), sumSession.steps.toString()),
					StatData(r.getString(R.string.stats_weekly_distance_travelled), "${sumSession.distanceInM / 1000} km")
			)))

			handleResponse(weeklyStats, AppendBehaviors.FirstFirst)


			val monthAgoCalendar = Calendar.getInstance()
			monthAgoCalendar.add(Calendar.MONTH, -1)
			//todo show all session for the past 24 hours and merge the rest
			sessionDao.getBetween(monthAgoCalendar.timeInMillis, now).forEach {

				//val locations = locationDao.getAllBetween(it.start, it.end)

				//val distance = calculateDistance(locations)

				val stats = arrayOf(Stat("${Assist.formatShortDateTime(it.start)} - ${Assist.formatShortDateTime(it.end)}", "", false, listOf(
						StatData(r.getString(R.string.stats_weekly_distance_travelled), "${(it.distanceInM / 1000.0).format(2)} km"),
						StatData(r.getString(R.string.stats_weekly_collected_location), Assist.formatNumber(it.collections)),
						StatData(r.getString(R.string.stats_weekly_steps), it.steps.toString())
				)))
				handleResponse(stats, AppendBehaviors.FirstLast)
			}


			/*val locations = AppDatabase.getAppDatabase(activity).locationDao().getAllSince(System.currentTimeMillis() - Constants.DAY_IN_MILLISECONDS * 30)
			locations.groupBy { it.location.time / Constants.DAY_IN_MILLISECONDS }.forEach {
				val distance = calculateDistance(it.value).format(2)
				//todo add localization support
				val stats = arrayOf(Stat(Assist.formatDate(activity, it.value.first().location.time), "", false, listOf(
						StatData("distance", "$distance km"),
						StatData("locations", Assist.formatNumber(it.value.size))
				)))
				GlobalScope.launch(Dispatchers.Main) {
					addStatsTable(stats, AppendBehaviors.Any)
					adapter!!.sort()
				}
			}*/

			GlobalScope.launch(Dispatchers.Main) {
				swipeRefreshLayout.isRefreshing = false
			}
		}
		//}
	}

	private fun calculateDistance(locations: List<DatabaseLocation>): Double {
		var totalDistance = 0.0
		val count = locations.size
		var last = locations[0]
		for (i in 1 until count) {
			val current = locations[i]

			if (current.location.time - last.location.time < Constants.MINUTE_IN_MILLISECONDS) {
				totalDistance += last.location.distanceFlat(current.location, LengthUnit.Kilometer)
			}

			last = current
		}

		return totalDistance
	}

	private fun handleResponse(value: Array<Stat>, @AppendBehaviors.AppendBehavior appendBehavior: Int) {
		GlobalScope.launch(Dispatchers.Main, CoroutineStart.DEFAULT) {
			addStatsTable(value, appendBehavior)
			adapter!!.sort()
		}
	}

	private fun generateMockData() {
		addStatsTable(generateMockStatList(), AppendBehaviors.Any)
	}

	private fun generateMockStatList(): Array<Stat> {
		val list = ArrayList<Stat>()
		for (i in 1..10) {
			list.add(generateMockStat(i))
		}
		return list.toTypedArray()
	}

	private fun generateMockStat(index: Int) = Stat("Mock $index", "donut", false, generateStatData(index))

	private fun generateStatData(index: Int): List<StatData> {
		val list = ArrayList<StatData>()
		for (i in 1..index) {
			list.add(StatData("Title $i", i.toString()))
		}
		return list
	}

	/**
	 * Generates tables from list of stats
	 *
	 * @param stats stats
	 */
	private fun addStatsTable(stats: Array<Stat>, @AppendBehaviors.AppendBehavior appendBehavior: Int) {
		for (s in stats) {
			val table = Table(s.data.size, s.showPosition, CARD_LIST_MARGIN, appendBehavior)
			table.title = s.name
			s.data.indices
					.asSequence()
					.map { s.data[it] }
					.forEach { table.addData(it.id, it.value) }
			adapter!!.add(table)
		}
	}


	override fun onEnter(activity: FragmentActivity) {

	}

	override fun onLeave(activity: FragmentActivity) {
	}


	override fun onPermissionResponse(requestCode: Int, success: Boolean) {

	}

	companion object {
		private const val CARD_LIST_MARGIN = 16
	}
}

